File: Algorithms/Python/22 Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def backtrack(left, right):
            if left + right == 2 * n:
                ans.append(''.join(parentheses))
                return

            if left < n:
                parentheses.append('(')
                backtrack(left + 1, right)
                parentheses.pop()
            
            if right < left:
                parentheses.append(')')
                backtrack(left, right + 1)
                parentheses.pop()

        ans = []
        parentheses = []
        backtrack(0, 0)
        return ans                

File: Algorithms/CPP/22 Generate Parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        this->n = n;
        backtrack(0, 0);
        return ans;
    }

    void backtrack(int left, int right) {
        if (left == n && right == n) {
            ans.push_back(parentheses);
            return;
        }

        if (left < n) {
            parentheses.push_back('(');
            backtrack(left + 1, right);
            parentheses.erase(parentheses.size() - 1);
        }

        if (right < left) {
            parentheses.push_back(')');
            backtrack(left, right + 1);
            parentheses.erase(parentheses.size() - 1);
        }
    }

private:
    string parentheses;
    vector<string> ans;
    int n;
};
